Fixed JBTM-2350 - CDI Transactional interceptors are not thread safe. (previousUserTransactionAvailability)

NiematojakTomasz opened the pull request:
!XTS !QA_JTA !QA_JTS_JACORB !BLACKTIE !PERF

Do you need unit tests for it? They seems easy to write, harder to launch `-Parq`.

mmusgrov commented:
To run the arq tests you need to set JBOSS_HOME:
export JBOSS_HOME=<wildfly dir>/build/target/wildfly-9.0.0.Alpha2-SNAPSHOT
./build.sh -f ./ArjunaJTA/cdi/pom.xml -Parq test

but to create a wildfly build that contains your fix to narayana you would need to run our build script:

export AS_BUILD=0 JTA_CDI_TESTS=1 OSGI_TESTS=0 NARAYANA_BUILD=0 NARAYANA_TESTS=0 BLACKTIE=0 XTS_AS_TESTS=0 XTS_TESTS=0 TXF_TESTS=0 txbridge=0 RTS_AS_TESTS=0 RTS_TESTS=0 QA_TESTS=0 SUN_ORB=0 JAC_ORB=1 JTA_AS_TESTS=0 PERF_TESTS=0

./scripts/hudson/narayana.sh

If you have unit tests then it would be nice to have them in case it breaks again in the future.

mmusgrov commented on ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/transactional/TransactionalInterceptorBase.java:
```diff
@@ -45,17 +46,36 @@
+        final TransactionManager tm = getTransactionManager();
+        final Transaction tx = tm.getTransaction();
+
+        boolean previousUserTransactionAvailability = setUserTransactionAvailable(false);
```
I would have expected you to use the value passed into the constructor (by the subclass) instead of false, as in:

boolean previousUserTransactionAvailability = setUserTransactionAvailable(userTransactionAvailable);

mmusgrov replied:
I think this explains why 2 of the CDI tests have just failed

mmusgrov commented:
So, apart from my two comments (using userTransactionAvailable in the base class and the question about unit tests), the fix looks good.

NiematojakTomasz commented:
@mmusgrov 
Fixed the bug you found. Tests passed on my PC.

I implemented 2 test suits - 18 test that have failed on previous implementation.
ConcurrentTransactionalTest - tests scenario described in issue, but with all combinations of tx types and transaction management types - 12 tests.
RecursiveTransactionalTest - tests recursive calls to intercepted method. Would still fail if previousUserTransactionAvailability would be thread local field.

Tests results before fix - https://gist.github.com/NiematojakTomasz/916da684904ce0d08dbd

tomjenkinson commented:
For info, the XTS failure looks to be due to something that has changed in upstream WFLY so its not an issue with the actual fix. Although as the tests failed we can't be certain a later test in the XTS suite would detect an issue in this PR.

NiematojakTomasz commented:
@tomjenkinson Should I rebase it or something?

tomjenkinson commented:
@NiematojakTomasz squashing it would be great please!

NiematojakTomasz commented:
@tomjenkinson Here you go.

mmusgrov commented:
That is a very nice set of tests. If you are happy with your PR then mark it with the label "Ready For Merge". If you have the rights then merge it yourself otherwise I will merge it for you.

Thank you very much for finding the bug and providing the fix and great tests.

NiematojakTomasz commented:
@mmusgrov I am happy with my pull request, but I don't have permission to label it (according to github help it requires write access - https://help.github.com/articles/permission-levels-for-an-organization-repository/). Or do you have some commands for bot to do it?

Please label it "Ready For Merge"  for me or merge. Thanks!

mmusgrov commented:
OK I have merged your changes. I will assign the JIRA to you so that you can mark it as resolved.

NiematojakTomasz commented:
@mmusgrov Should I also close it as a reporter?

mmusgrov commented:
No. Our process is that the assigned developer marks it as resolved once it has been merged. Later when we come to do a release that contains the fix the release engineer will mark it as closed. Your fix is targeted towards the next release which is 5.0.5

tomjenkinson commented:
And thank you very much for your contribution!!!

mmusgrov commented:
OK Tomasz I have added you as a contributor and assigned you the JIRA. Use the Resolve Issue button to mark it as fixed.
